lib/evmone/eof.cpp

```diff
@@ -34,51 +61,85 @@ std::variant<EOFSectionHeaders, EOFValidationError> validate_eof_headers(
+    // TODO: Since all sections are mandatory and they have to be ordered (Types, Code+, Data)
+    // TODO: this fragment of code can be much simpler. Rewriting needed.
+    uint8_t expected_section_id = TYPE_SECTION;
     while (it != container_end && state != State::terminated)
     {
         switch (state)
         {
         case State::section_id:
         {
             section_id = *it++;
+
+            if (section_id != expected_section_id)
+                return get_section_missing_error(expected_section_id);
+
             switch (section_id)
             {
             case TERMINATOR:
-                if (section_headers[CODE_SECTION] == 0)
-                    return EOFValidationError::code_section_missing;
                 state = State::terminated;
                 break;
-            case DATA_SECTION:
-                if (section_headers[CODE_SECTION] == 0)
-                    return EOFValidationError::code_section_missing;
-                if (section_headers[DATA_SECTION] != 0)
-                    return EOFValidationError::multiple_data_sections;
+            case TYPE_SECTION:
+                expected_section_id = CODE_SECTION;
                 state = State::section_size;
                 break;
             case CODE_SECTION:
-                if (section_headers[CODE_SECTION] != 0)
-                    return EOFValidationError::multiple_code_sections;
+            {
+                if (it >= container_end - 2)
+                    return EOFValidationError::incomplete_section_number;
+                section_num = read_uint16_be(it);
+                it += 2;
+                if (section_num == 0)
+                    return EOFValidationError::zero_section_size;
+                expected_section_id = DATA_SECTION;
+                state = State::section_size;
+                break;
+            }
+            case DATA_SECTION:
+                expected_section_id = TERMINATOR;
                 state = State::section_size;
                 break;
             default:
-                return EOFValidationError::unknown_section_id;
+                assert(false);
             }
             break;
         }
         case State::section_size:
         {
-            const auto size_hi = *it++;
-            if (it == container_end)
-                return EOFValidationError::incomplete_section_size;
-            const auto size_lo = *it++;
-            const auto section_size = static_cast<uint16_t>((size_hi << 8) | size_lo);
-            if (section_size == 0)
-                return EOFValidationError::zero_section_size;
-
-            section_headers[section_id] = section_size;
+            if (section_id == CODE_SECTION)
+            {
+                assert(section_num > 0);  // Guaranteed by previous validation step.
+                for (size_t i = 0; i < section_num; ++i)
+                {
+                    if (it >= container_end - 2)
```

Comment:
`it` here points to the next byte after code_section_num (first byte of single section size), I think the right condition here is
```suggestion
                if (it + 1 >= container_end)
```

Reply:
Fixed.